src/python/WMArchive/PySpark/RecordAggregator.py

```diff
@@ -14,71 +27,125 @@ def mapper(self, records):
+
+                # Performance
+                if not stats.get('performance'):
+                    stats['performance'] = {
+                        "cpu": {
+                            "TotalJobCPU": 0,
+                            "TotalJobTime": 0,
+                        },
+                        "storage": {
+                            "readTotalMB": 0,
+                            "writeTotalMB": 0,
+                        }
+                    }
+                performance_cpu = step['performance']['cpu']
+                if performance_cpu.get('TotalJobCPU', 0):
+                    stats['performance']['cpu']['TotalJobCPU'] += performance_cpu.get('TotalJobCPU', 0)
+                if performance_cpu.get('TotalJobTime', 0):
+                    stats['performance']['cpu']['TotalJobTime'] += performance_cpu.get('TotalJobTime', 0)
+                performance_storage = step['performance']['storage']
+                if performance_storage.get('readTotalMB', 0):
+                    stats['performance']['storage']['readTotalMB'] += performance_storage.get('readTotalMB', 0)
+                if performance_storage.get('writeTotalMB', 0):
+                    stats['performance']['storage']['writeTotalMB'] += performance_storage.get('writeTotalMB', 0)
+
+                # Store stats in document
+                document['stats'][scope_hash] = stats
+
+        return document
 
     def reducer(self, records, init=0):
         "Simpler reducer which collects all results from RDD.collect() records"
-        out = {}
-        count = 0
-        mdict = {'cpu':0, 'time':0, 'rsize':0, 'wsize':0, 'doc':0}
-        for rec in records:
-            for host, hdict in rec.items():
-                if  host not in out.keys():
-                    out[host] = {}
-                _hdict = out[host]
-                for site, sdict in hdict.items():
-                    if  site not in _hdict.keys():
-                        _hdict = {site:mdict}
-                        out[host][site] = mdict
-                    _sdict = _hdict[site]
-                    _cpu = _sdict['cpu'] + sdict['cpu']
-                    _time = _sdict['time'] + sdict['time']
-                    _rsize = _sdict['rsize'] + sdict['rsize']
-                    _wsize = _sdict['wsize'] + sdict['wsize']
-                    _doc = _sdict['doc'] + sdict['doc']
-                    ndict = {'cpu':_cpu, 'time':_time, 'rsize':_rsize, 'wsize':_wsize, 'doc':_doc}
-                    out[host][site] = ndict
-            count += 1
-	# simple printout of summary info
-        for host, hdict in out.items():
-            print('\n### %s' % host)
-            for site, sdict in hdict.items():
-                print('')
-                print(site)
-                for key, val in sdict.items():
-                    print('%s: %s' % (key, val))
-        return out
+        document = {
+            'stats': {},
+        }
+        for existing_document in records:
```

Comment:
ok,
few things first:

- you should not create a new MongoClient in reducer, instead created once in
  constructor of MapReduce class
```
  def __init__(self):
     self.client = MongoClient()
     self.coll = # your collection
     self.chunk_size = 1000 # a chunk size for bulk insert
```

- your reducer takes records so you can pass them into stand-alone function
  which will modify them and yield them back, e.g.

```
def prepare_records(records):
    for existing_document in records:
        # do stuff
        for scope_hash, existing_stats in existing_document['stats'].items():
            # do sutff
            yield stats

```

now modify your reducer method to use prepare_records generator function:

```
import pymongo # to check pymongo version for its APIs
import itertools # to use iterator over generator records
def reducer(self, records, init=0):
    gen = prepare_records(records)
	size = self.cache_size
	try:
		if  pymongo.version.startswith('3.'): # pymongo 3.X
			res = self.coll.insert_many(gen)
			inserted += len(res.inserted_ids)
		else:
			while True:
				nres = self.coll.insert(itertools.islice(gen, size))
				if  nres and isinstance(nres, list):
					inserted += len(nres)
				else:
					break
	except InvalidDocument as exp:
       # handle exceptions
```

This code refactoring benefit from:
- has only single MongoClient instance, and reducer will re-use it
- delegate records manipulate into stand-along function (which you can test as you like with any records types)
- generate records via generator and pass generator to mongo collection self.coll
- handle mongo insert exceptions
- don't need to write json files
- work with pymongo 2.X and 3.X APIs to insert bulk of records into collection

Please re-factor code in similar fashion, test it and verify that you insert new
records into MongoDB.


On  0, Nils Fischer <notifications@github.com> wrote:
> > -                    ndict = {'cpu':_cpu, 'time':_time, 'rsize':_rsize, 'wsize':_wsize, 'doc':_doc}
> > -                    out[host][site] = ndict
> > -            count += 1
> > -	# simple printout of summary info
> > -        for host, hdict in out.items():
> > -            print('\n### %s' % host)
> > -            for site, sdict in hdict.items():
> > -                print('')
> > -                print(site)
> > -                for key, val in sdict.items():
> > -                    print('%s: %s' % (key, val))
> > -        return out
> > +        document = {
> > +            'stats': {},
> > +        }
> > +        for existing_document in records:
> 
> So in my code the mapper fills the `document` dictionaries and the reducer merges them. It's only a dictionary to have an efficient lookup mechanism via the hashed scope. After the merging I grab all of the dictionary's values and insert them into MongoDB. That's now a list of documents, not one huge document anymore.
> 
> Since the reducer first needs to do the merging I am not sure how to insert chunks into MongoDB before.
> 
> ---
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMArchive/pull/205/files/b7826071b0059bfde9c1b0f24962fc0eeb89220f#r73369640